nav2_collision_monitor/README.md

```diff
@@ -0,0 +1,61 @@
+* Robot footprint polygon, which is used in the approach behavior model only.
+* Circle, used in all models: for stop and slowdown models as a safety area, for approach model as a robot footprint. Circle is made for the best performance and could be used in the cases where the safety area or robot could be approximated by round shape.
+
+NOTE: Although safety behavior models are not intended to be used simultaneously (e.g. stop model should not be crossed with approach one), it is not prohibited to. Collision Monitor allows setting simultaneously multiple shapes with different behavior models. This typically could be useful to have a small stop area and larger slowdown bounding box to warn the robot from a collision without termination of operation.
+
+The obstacle points are being obtained from different data sources. Collision Monitor is subscribed to:
+
+* Laser scanners (`sensor_msgs::msg::LaserScan` messages)
+* PointClouds (`sensor_msgs::msg::PointCloud2` messages)
+* IR/Sonars (`sensor_msgs::msg::Range` messages, not implemented yet)
+
+## Design
+
+Since Collision Monitor is designed to operate as an independent safety node, Nav2 stack has no knowledge about it.
+It is laying under the stack and operating after all necessary decisions were made by Nav2.
+This is achieved through remapped `cmd_vel` topic, going out from a Controller.
+
+The following diagram is showing the high-level design of Collision Monitor module. All shapes (Polygons and Circles) are derived from base `Polygon` class, so without loss of generality we can call them as polygons. Subscribed footprint is also having the same properties as other polygons, but it is being obtained from `nav2_costmap_2d::FootprintSubscriber`.
+![HDL.png](doc/HLD.png)
+
+## Configuration
+
+Detailed configuration parameters, their description and how to setup a Collision Monitor could be found at its [Configuration Guide](https://navigation.ros.org/configuration/packages/configuring-collision-monitor.html) and [Using Collision Monitor tutorial](https://navigation.ros.org/tutorials/docs/using_collision_monitor.html) pages.
+
+
+## Metrics
+
+Designed to be used in wide variety of robots (incl. moving fast) and have a high level of reliability, Collision Monitor node should operate at fast rates.
+Typical one frame processing time is ~4-5ms for laser scanner (with 360 points) and ~4-80ms for PointClouds (having 24K points).
+The table below represents the details of operating times for different behavior models and shapes:
+
+| | Stop/Slowdown model, Polygon area | Stop/Slowdown model, Circle area | Approach model, Polygon footprint | Approach model, Circle footprint |
+|-|-----------------------------------|----------------------------------|-----------------------------------|----------------------------------|
+| LaserScan (360 points) processing time, ms  | 4.09 | 4.08 | 4.98  | 4.29  |
+| PointCloud (24K points) processing time, ms | 4.13 | 3.76 | 77.92 | 11.43 |
```

Comment:
Is there a way to appreciably speed up the approach model polygon? Maybe a different algorithm or algorithm optimizations?

Reply:
Try to embed the `getPointInside` inside of `getPointsInside`?

Reply:
Multithreading?

Reply:
Yes, I thought about ways of optimization.

[Here](https://erich.realtimerendering.com/ptinpoly/) is the work of comparing the performance for different algorithms about point-in-polygon solution. Among basic algorithms, raycrossing algorithm is the fastest one. There is an optimization, called as "MacMartin" used for large polygons, which speeds up the raycrossing algorithms on x1.8 times.
The best way of enhance algorithms: is a grid algorithm, quantizing the map and making "inside", "non-inside" and "indeterminate" cells. According to the work, this algorithm gives best performance, and thus more promising as an optimization point.

> Try to embed the getPointInside inside of getPointsInside?

This function was already inlined locally: `inline bool Polygon::isPointInside(const Point & point) const`, and confirmed by binary disassembling.

The multithreading is another way of optimization. Currently, the motion is being simulated by small steps. But it could being calculated the exact position of robot pose after `t` time (since we have `theta` and linear velocity, this making arcs; and and then we need to make a transform from robot's velocity frame down to base frame). Knowing the exact robot position in any time frame, we could parallel the whole calculation loop e.g. by giving odd iterations to one thread, and even ones - to another. All iterations will be independent from each other. This is another point of optimization could be made there.

Reply:
Update: the multithreading could be parallelized by points, instead of simulation steps. This makes it easiest to implement giving tangible benefits on multi-CPU systems.

Reply:
What would a default of `0.1` for simulation timestep yield? I think 0.1s timestep increments are plenty small and are 5x larger than currently set.

Reply:
That is definitely makes sense. The Footprint Approach time was decreased from ~78 -> ~20 ms. Re-measured the peroformance and [updated](https://github.com/ros-planning/navigation2/pull/2982/commits/a4639aac3380a3bdbf918895319cb0e2721f680f) the README accordingly.